quant: add QAT fused Linear-Bn1d [1/x]: prepared module

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72796
* __->__ #72431

Summary:

Adds support for a fused QAT observed module for `Linear` followed by
`BatchNorm1d`. In this PR, only the support for prepared module with
fake_quants in the right places is added.

A future PR will add support for `convert`, and tests for eager and FX
graph mode workflows.

Similar to conv-bn, we rescale the weight before applying the fake
quant, and undo the rescaling after the linear operation.

Test Plan:

```
python test/test_quantization.py TestQuantizeEagerQATNumerics.test_linear_bn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34044427](https://our.internmc.facebook.com/intern/diff/D34044427)

vkuzo committed:
quant: add QAT fused Linear-Bn1d [1/x]: prepared module

Summary:

Adds support for a fused QAT observed module for `Linear` followed by
`BatchNorm1d`. In this PR, only the support for prepared module with
fake_quants in the right places is added.

A future PR will add support for `convert`, and tests for eager and FX
graph mode workflows.

Similar to conv-bn, we rescale the weight before applying the fake
quant, and undo the rescaling after the linear operation.

Test Plan:

```
python test/test_quantization.py TestQuantizeEagerQATNumerics.test_linear_bn
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34044427).

jerryzh168 commented on torch/nn/intrinsic/qat/modules/linear_fused.py:
```diff
@@ -0,0 +1,154 @@
+from torch.nn.parameter import Parameter
+
+
+class LinearBn1d(nn.modules.linear.Linear, nni._FusedModule):
```
nit: nn.modules.linear.Linear can be nn.Linear?

jerryzh168 commented on torch/nn/intrinsic/qat/modules/linear_fused.py:
```diff
@@ -0,0 +1,154 @@
+                 # Args for this module
+                 freeze_bn=False,
+                 qconfig=None):
+        nn.modules.linear.Linear.__init__(self, in_features, out_features, bias)
```
same here

vkuzo commented on torch/nn/intrinsic/qat/modules/linear_fused.py:
```diff
@@ -0,0 +1,154 @@
+        #   x1_orig = x1_scaled / scale_factor + b
+        #   x2 = self.bn(x1_orig)
+
+        running_std = torch.sqrt(self.bn.running_var + self.bn.eps)
```
this is the same logic as fused QAT conv-bn

jerryzh168 replied:
oh OK, sounds good

vkuzo committed:
Update on "quant: add QAT fused Linear-Bn1d [1/x]: prepared module"


Summary:

Adds support for a fused QAT observed module for `Linear` followed by
`BatchNorm1d`. In this PR, only the support for prepared module with
fake_quants in the right places is added.

A future PR will add support for `convert`, and tests for eager and FX
graph mode workflows.

Similar to conv-bn, we rescale the weight before applying the fake
quant, and undo the rescaling after the linear operation.

Test Plan:

```
python test/test_quantization.py TestQuantizeEagerQATNumerics.test_linear_bn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34044427](https://our.internmc.facebook.com/intern/diff/D34044427)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34044427).

vkuzo committed:
Update on "quant: add QAT fused Linear-Bn1d [1/x]: prepared module"


Summary:

Adds support for a fused QAT observed module for `Linear` followed by
`BatchNorm1d`. In this PR, only the support for prepared module with
fake_quants in the right places is added.

A future PR will add support for `convert`, and tests for eager and FX
graph mode workflows.

Similar to conv-bn, we rescale the weight before applying the fake
quant, and undo the rescaling after the linear operation.

Test Plan:

```
python test/test_quantization.py TestQuantizeEagerQATNumerics.test_linear_bn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34044427](https://our.internmc.facebook.com/intern/diff/D34044427)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34044427).
